### PR TITLE
Node.js tool for detecting Entries that share 99% of their area (Duplicates)

### PR DESCRIPTION
**Pull request by DRKSLV**
Hi, i made a handy tool for getting a list of duplicates

Assigns duplicates based on Polygons: `intersec_area/outer_area < 0.01`
Almost no false positives, quite a few false negatives when borders of duplicate are drawn sloppily

The script also generates a second version of the atlas:
Adds parent: "" property to the atlas, that lists the id of the first identical element
Adds bounds: [] property with bounding box
Adds childs: [] property with the ids of the children
(These properties could be used to fold all duplicates in a comment section style below the parent entry, so they don't take up space)

To run, download Node.js and run `node tools/duplicateFinder`

### CODE REVIEW

**Comment by DRKSLV** [comment]
Oop, i'll have to fix this tomorrow, but it seems that the vertex order of the intersected polygon is wrong, and could throw off area calculations
This is because i accidentally implemented the convex version of the intersection algorithm
The concave one requires an additional step to determine the vertex order

Also in line 24 i call orderClockwise on the input polygon
i think that again has the catch of messing up concave polygons 
(that only throws off the area calculation, as the modified polygon isn't written to the file, but its definetely bad)

so i'll have to fix some of that stuff tomorrow, sorry guys :(

File: web/_js/pointInPolygon.js [hunk 1]
```diff
@@ -45,3 +45,6 @@ function pointIsInPolygon (point, polygon) {
     
     return inside;
 };
+
+if( typeof window == "undefined") // for use in /tools using nodejs
+    module.exports = pointIsInPolygon;
```
Comment (nathanfranke) [inline web/_js/pointInPolygon.js:50]:
```suggestion

// for use in node.js tooling
if (typeof window == "undefined") {
    module.exports = pointIsInPolygon;
}
```

Include a trailing newline to be [POSIX compliant](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline).

File: tools/duplicateFinder.js [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+let isec = intersectPolygons( tri,slab);
+console.log( isec )
+console.log( polygon_area( isec) )
+*/
+/*
+console.log( doBoundsCollide( [ 0, 0, 2, 2 ], [1, 1, 2, 2] )  )
+console.log( doBoundsCollide( [ 0, 0, 2, 2 ], [-1, -1, 2, 2] ))
+console.log( doBoundsCollide( [ 0, 0, 2, 2 ], [2.01, 0, 2, 2] ))
+console.log( doBoundsCollide( [ 0, 0, 2, 2 ], [0, 2.01, 2, 2] ))
+console.log( doBoundsCollide( [ 0, 0, 2, 2 ], [0, -2.01, 2, 2] ))*/
```
Comment (nathanfranke) [inline tools/duplicateFinder.js:216]: Include a trailing newline to be [POSIX compliant](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline).

**Review by nathanfranke — CHANGES_REQUESTED** [review]
Note: I am not a maintainer of this repository.

File: web/_js/infoblock.js [hunk 3]
```diff
@@ -41,7 +41,7 @@ function createInfoBlock(entry) {
     }
     let idElement = document.createElement("p");
     idElement.style.fontFamily = "Dejavu Sans Mono, sans, Sans-Serif;";
-    idElement.innerText = "id: " + entry.id;
```
[the remark this answers is not in the document]
Comment (DRKSLV) [reply web/_js/infoblock.js:44]: I had previously had a change here and removed it
A semicolon was left over from my tampering

I can remove it for clarity but the line has to stay

**Comment by DRKSLV** [comment]
Hi, good and bad news
i had a little less time today,
and unfortunately hit a 2h+ dead end trying to understand the "Weiler Atherton"-Algorithm.
Mostly because i was learning from the ramblings of a paranoid scizophrenic on stack overflow with 0 upvotes,
instead of taking the 5minutes to find an actual resource...

Good news: I am on track now and it's a lot less complicated than SO makes it seem, so i'm probably going to reach an acceptable state soon, where ppl can already use the tool to manually find duplicates, and then it may or may not be integrated further, depending on the projects needs and yalls opinion

**Comment by DRKSLV** [comment]
Currently there is a problem with edge cases, when points of one polygon are placed exactly on top of lines of the other polygon
This is unfortunately very common and results in errors such as the GME line being classified as a duplicate of the "dutch right region", due to the intersection polygon being way larger than it should be
This is obviously unacceptable, an i'll try to work on it further tomorrow

Apparently, intersection points need to be classified by "entering", "exiting" and "not actually an intersection lmao this is cursed"
Currently, this just causes broken polygons

I'll try to get to it tomorrow

**Comment by DRKSLV** [comment]
Ah, and heads up!
The code is currently not even doing it's main task, instead it's running tests, that i manually comment and uncomment

This will all be removed if and when this is ready to merge

**Comment by ProstoSanja** [comment]
Can you please mark the PR as draft until its done?

**Comment by AnonymousRandomPerson** [comment]
This appears to be inactive. I'll close this PR; feel free to reopen if you'd still like to contribute this.

**Comment by DRKSLV** [comment]
Fair: sorry i really got overwhelmed with some unexpected stuff

Maybe ill resume work on this later

**Pull request closed without merging**